sparse/src/KokkosSparse_coo2crs.hpp

```diff
@@ -0,0 +1,352 @@
+    auto cpy_end = m_crs_row_map(row_idx + 1);
+    auto cpy_len = cpy_end - cpy_beg;
+
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(member, 0, cpy_len),
+                         [&](const CrsOT &i) {
+                           auto offset           = i + cpy_beg;
+                           m_crs_vals(offset)    = m_umaps[i].value_at(i);
+                           m_crs_col_ids(offset) = m_umaps[i].key_at(i);
+                         });
+  }
+
+  Coo2Crs(DimType m, DimType n, RowViewType row, ColViewType col,
+          DataViewType data) {
+    m_n_tuples = data.extent(0);
+    m_nrows    = m;
+    m_ncols    = n;
+    m_row      = row;
+    m_col      = col;
+    m_data     = data;
+
+    typename UmapType::size_type arg_capacity_hint =
+        m_nrows > 0 ? (m_n_tuples / m_nrows / 4) : 16;
+    typename UmapType::hasher_type arg_hasher;
+    typename UmapType::equal_to_type arg_equal_to;
+    arg_capacity_hint = arg_capacity_hint < 16 ? 16 : arg_capacity_hint;
+
+    // Record of whether capacity was reached in any unordered map
+    m_capacity_bmap = BmapViewType("m_capacity_bmap", m_nrows);
+    typename BmapViewType::HostMirror m_capacity_bmap_mirror =
+        Kokkos::create_mirror_view(m_capacity_bmap);
+
+    // Track which tuples have been processed
+    m_tuple_bmap = Bitset(m_n_tuples);
+
+    m_crs_row_map = CrsRowMapView(
+        Kokkos::view_alloc(Kokkos::WithoutInitializing, "m_crs_row_map"),
+        m_nrows + 1);
+
+    // Memory management notes for `umap_ptrs` and `m_umaps`:
+    //   `umap_ptrs` is a two dimensional array. The first dimension contains
+    //   pointers to mixed-memory (host and device memory). The second
+    //   dimension is the array of UnorderedMap objects. Some of the object
+    //   methods are callable from only the device (device-callable), others
+    //   are callable from only the host. Some of the host-callable methods,
+    //   such as rehash are intended to be observable on the device.
+    //   See Kokkos::UnorderedMap for details.
+    //
+    //   `m_umaps` is a single dimension array of device memory. This array
+    //   contains a shallow copy of all the UnorderedMap members that are
+    //   allocated manually below.
+    //
+    //   Any time a host-callable method with device observable results is
+    //   invoked, we must shallow-copy the given `umap_ptrs` member back to
+    //   the device.
+    //
+    //   However, since we are using shallow copies of objects of type
+    //   UnorderedMap, we do not need to copy the device memory back to
+    //   the host before using a host-callable method.
+
+    // Setup a nrows length array of Unordered Maps
+    m_umaps =
+        reinterpret_cast<UmapType *>(Kokkos::kokkos_malloc<UmapMemorySpace>(
+            "m_umaps", m_nrows * sizeof(UmapType)));
+
+    using shallow_copy_to_device =
+        Kokkos::Impl::DeepCopy<UmapMemorySpace, typename Kokkos::HostSpace>;
+
+    UmapType **umap_ptrs = new UmapType *[m_nrows];
```

Comment:
Did I understand this right? Are we setting up nrows length arrays nrow times? That is a dense matrix!

Reply:
Not quite. We are setting up nrows unordered maps -- 1 unordered map for each row. Each unordered map's initial capacity is >= 16.